[WIP] Coproc stuff3

otyshchenko1 opened the pull request:
Please find free time to review patch series.

otyshchenko1 committed:
arm/coproc: Add initial draft for dummy remote processor coproc_xxx

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
arm/coproc: Add initial draft for dummy Round-Robin scheduler

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
arm/coproc: First draft of generic remote processors framework

Both coproc core and coproc scheduler are integrated into
the framework.

The patch also contains functional for parsing coproc configuration
for dom0 and passing needed coproc to dom0 device tree.

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
libxl/coproc: Introduce coproc configuration for domU

Sync coproc devices that are described in domain config file and
coproc devices that will be copied to guest domain
device tree.

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

klogg commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,422 @@
+
+    list_for_each_entry(coproc, &coprocs, coproc_elem)
+    {
+        if ( !strcmp(dev_path(coproc->dev), path) )
```
strncmp() everywhere, please

otyshchenko1 replied:
I will change

klogg commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,320 @@
+	sched->sched_priv = sched_data;
+	init_timer(&sched_data->s_timer, s_timer_fn, sched, 0);
+	vcoproc_scheduler_set_current(sched, NULL);
+	spin_lock_init(&sched_data->schedule_lock);
```
More spinlocks for the spinlocks god! I am scared of races, deadlocks, maintenance issues

otyshchenko1 replied:
I got your point here. I will think about it.

otyshchenko1 committed:
arm/coproc: temp: Debug coproc-emu kernel driver operations

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

andr2000 commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,381 @@
+    {
+        spin_unlock(&vcoproc_d->lock);
+
+        ret = coproc_detach_from_domain(d, vcoproc);
```
is it required that coproc_detach_from_domain holds the lock?
can it be w/o lock and users will lock themselves?

otyshchenko1 replied:
I will think about how to optimize.

andr2000 commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+		vcoproc->state = VCOPROC_SLEEPING;
+	else
+	{
+		vcoproc->state = VCOPROC_TERMINATING;
```
this is not obvious why going to state TERMINATING if asked to sleep

otyshchenko1 replied:
VCOPROC_TERMINATING state means that currently running vcoproc (I mean vcoproc instance that currently runs on real coproc device) was asked to sleep. The running at the moment vcoproc might not go to sleep immediately as we want, so we can't set VCOPROC_SLEEPING state for it. 

We have already come to conclusion about using more appropriate name for it -> VCOPROC_ASKED_TO_SLEEP to not confuse anyone). So, I will change

andr2000 commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+	int reschedule = 0;
```
bool?

otyshchenko1 replied:
agree

andr2000 commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+		vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_yield(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
```
don't you need a return value?
or you don't care/not needed to know if actually was able to yield?

otyshchenko1 replied:
From my point of view the yield function like wake/sleep doesn't need a return value, since state is being changed inside function. At least until I won't see "the whole picture".

andr2000 commented:
My feeling is that we shouldn't put it under ARM as it was intended to be generic and it seem to be

andr2000 commented:
For tests I would recommend we implement a timer co-proc:
1. Has real interrupts
2. Has real registers
3. Can be relatively easy implemented/tested

aanisov commented:
xen/drivers could be a good place for the stuff. F.e. smmu support which is pure ARM stuff you could find in xen/drivers/passthrough/arm, and is an instance of passthrough functionality.

aanisov commented:
BTW, it would be great to have structures and functions documented in header files

aanisov commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,381 @@
+        return 0;
+
+    /* It is no matter who will provide coproc */
+    coproc = next ? next->coproc : curr->coproc;
```
The comment above is not really clear. I would suggest you to remove it or to make it more obvious.

otyshchenko1 replied:
Will do

aanisov commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,381 @@
+    if ( !coproc )
+        return -EINVAL;
+
+    spin_lock(&coprocs_lock);
```
You have no need to take a coproc list spinlock here. It is not an operation with the coprocs list. You are instantiating a domains vcoproc for a particular coproc existing within the system.

otyshchenko1 replied:
The coprocs_lock protects not only the list but other operations with the coproc too.

aanisov commented:
I'm not sure finally, but it seems you do not need nested locks on _vcoproc_d->lock_ and _coprocs_lock_.

aanisov commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,381 @@
+    return coproc->ops->vcoproc_is_created(d, coproc);
+}
+
+static int vcoproc_dom0_init(struct domain *d)
```
__init

otyshchenko1 replied:
Agree. Will do

aanisov commented on xen/arch/arm/coproc/coproc.c:
```diff
@@ -0,0 +1,381 @@
+    return ret;
+}
+
+bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
```
__init, it is used in dom0 creation path only

otyshchenko1 replied:
Currently, yes. But I am going to use it for other domain too. So, nothing need to be changed.

aanisov commented on xen/arch/arm/coproc/coproc.h:
```diff
@@ -0,0 +1,110 @@
+struct vcoproc_ops {
+    struct vcoproc_instance *(*vcoproc_init)(struct domain *, struct coproc_device *);
+    void (*vcoproc_free)(struct domain *, struct vcoproc_instance *);
+    bool_t (*vcoproc_is_created)(struct domain *, struct coproc_device *);
```
is_created isn't a platform specific operation

otyshchenko1 replied:
I know, will do as soon as I move common parts to the framework

aanisov replied:
ok

aanisov commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+	if ( unlikely(!curr && !next) )
+		goto out;
+
+	if ( vcoproc_context_switch(curr, next) )
```
vcoproc_context_switch return value could be a good way to get the "time to wait" originated from the platform specific coprocessor internals:

- 0 - context switched

- x!=0  - wait for x milliseconds

otyshchenko1 replied:
Agree. Will do

aanisov commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+		if ( coproc_wait_time >= 0 )
+			set_timer(&sched_data->s_timer, now + coproc_wait_time);
+
+		VCOPROC_SCHED_OP(sched, schedule_completed, next, 0);
```
Why "next"? We failed to switch to it.

otyshchenko1 replied:
Yes, we failed to switch to "next". So, we have to put "next" back to the runqueue head.
For doing this we have to pass "0" thru last param.

aanisov replied:
ack it

aanisov commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+		return true;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	destroyed = vcoproc->state == VCOPROC_UNKNOWN ? true : false;
```
why do you need a spinlock here?

otyshchenko1 replied:
to protect vcoproc->state

aanisov replied:
You just read the current value of the variable, do not change it or any associated to it resources. There is no need to atomize pure read operations.

aanisov replied:
Well, this makes sense only in case if you associate _sched_data->schedule_lock_ with _vcoproc_. But then you should put _if (!vcoproc)_ under the spinlock as well.

otyshchenko1 replied:
Agree to remove lock here

aanisov commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	vcoproc->sched_priv = VCOPROC_SCHED_OP(sched, alloc_vdata, vcoproc);
+	if ( !vcoproc->sched_priv )
```
why do you need to take a spinlock on data allocation?

otyshchenko1 replied:
Taking a spinlock during data allocation is not needed _for now_. But I would stick to protect it too and make both data allocation and state changing atomic. Also you may add some code in future to the alloc_vdata that works with runqueue (or another resource that should be protected too) but forgetting to protect it here you may get into the trouble.

aanisov commented on xen/arch/arm/coproc/schedule.c:
```diff
@@ -0,0 +1,317 @@
+		return -ENOMEM;
+	}
+
+	vcoproc->state = VCOPROC_SLEEPING;
```
vcoproc->state setting seems to be the only thing here to be protected by spinlock, if any

otyshchenko1 replied:
See my previous comment

aanisov commented:
As we discussed last Friday, I would suggest to drop the functionality introduced by commit [libxl/coproc: Introduce coproc configuration for domU](https://github.com/aanisov/xen/pull/6/commits/cae331700ad9a076a7af7d19856b5613ff0751e8). 
At this moment I guess it would be great to stay with the Julien's passthrough configuration, just introducing an alternative way to handle configuration (assign shared resources) in hypervisor for resources shared by the framework.

otyshchenko1 committed:
arm/irq: Reorder check when the IRQ is already used by someone

Call irq_get_domain for the IRQ we are interested in
only after making sure that it is the guest IRQ to avoid
ASSERT(test_bit(_IRQ_GUEST, &desc->status)) triggering.

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
arm/coproc: Give permission to the guest to manage coproc irqs

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
flask: Add XEN_DOMCTL_attach_coproc support

And allow the source domain access to attach_coproc
while creating domain.

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
arm32_defconfig: Enable XSM support

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
flask: Allow a irq_t device to be used by a guest domain

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
arm/irq: Allow the IRQ to be shared between Xen and domains

At first this is needed to for the coproc irqs that should be owned by
XEN and used in domains where coproc is required.

This patch should be reworked properly taking into account that not all
XEN irqs are coproc irqs.

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 commented:
@aanisov said:
_As we discussed last Friday, I would suggest to drop the functionality introduced by commit libxl/coproc: Introduce coproc configuration for domU.
At this moment I guess it would be great to stay with the Julien's passthrough configuration, just introducing an alternative way to handle configuration (assign shared resources) in hypervisor for resources shared by the framework._

In general I agree with you. But, I would leave it as is at least until Julien has a look at this patch. Maybe he will have another opinion due to significant difference between "passthrough device" and "share device".

otyshchenko1 commented:
@aanisov said:
_xen/drivers could be a good place for the stuff. F.e. smmu support which is pure ARM stuff you could find in xen/drivers/passthrough/arm, and is an instance of passthrough functionality._

I don't mind to put coproc here. But later as we already have found common ground about it.

otyshchenko1 commented:
@aanisov said:
_BTW, it would be great to have structures and functions documented in header files_

Agree, my fault

aanisov commented:
BTW, I guess three(or four if we keep config stuff) first patches are the good base and should not be forcibly replaced. Please progress further with separate patches for comments addressing and development.
At the moment we decide to throw the stuff into mailing list we would squash fixes, but, IMHO not earlier.

otyshchenko1 commented:
@aanisov said:
_BTW, I guess three(or four if we keep config stuff) first patches are the good base and should not be forcibly replaced. Please progress further with separate patches for comments addressing and development.
At the moment we decide to throw the stuff into mailing list we would squash fixes, but, IMHO not earlier._

Agree, all comments here for these first 4 patches will be addressed and squashed with them. Other patches will be separate patches like new 6 patches I have pushed recently.

otyshchenko1 commented:
@aanisov said:
_I'm not sure finally, but it seems you do not need nested locks on vcoproc_d->lock and coprocs_lock._

I will rework coproc.c to use only one coprocs_lock. These will address both you and @andr2000 comments.

aanisov commented:
I guess the patch [arm/irq: Reorder check when the IRQ is already used by someone](https://github.com/aanisov/xen/pull/6/commits/bd97ad5e7773c0c3f694c8fb324fd118b1f2aee4) is pretty ready to be pushed to the mailing list.

aanisov commented on xen/arch/arm/domain_build.c:
```diff
@@ -1102,6 +1102,60 @@ static int map_device_children(struct domain *d,
+        /*
+         * Don't map IRQ that have no physical meaning
+         * ie: IRQ whose controller is not the GIC
+         */
```
copy-pasted comment

aanisov commented on xen/arch/arm/domain_build.c:
```diff
@@ -1102,6 +1102,60 @@ static int map_device_children(struct domain *d,
+            dt_dprintk("irq %u not connected to primary controller. Connected to %s\n",
+                      i, dt_node_full_name(rirq.controller));
+            continue;
+        }
```
Piece of copy-pasted code seems to be not needed here.

aanisov commented:
Patch [flask: Add XEN_DOMCTL_attach_coproc support](https://github.com/aanisov/xen/pull/6/commits/9134ba49fabbaa4920fdc87194bc64b83577681d) would be needed only in case we keep patch [libxl/coproc: Introduce coproc configuration for domU
](https://github.com/aanisov/xen/pull/6/commits/cae331700ad9a076a7af7d19856b5613ff0751e8)

otyshchenko1 committed:
Address comments for generic framework

Should be squashed with
arm/coproc: First draft of generic remote processors framework

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
Address comments for coproc_xxx platform specific code

Should be squashed with
arm/coproc: Add initial draft for dummy remote processor coproc_xxx

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
Address comments for Round-Robin scheduling algorithm

Should be squashed with
arm/coproc: Add initial draft for dummy Round-Robin scheduler

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
Code style fixes for Round-Robin scheduling algorithm

Should be squashed with
arm/coproc: Add initial draft for dummy Round-Robin scheduler

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
Code style fixes for generic framework

Should be squashed with
arm/coproc: First draft of generic remote processors framework

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 committed:
Code style fixes for coproc_xxx platform specific code

Should be squashed with
arm/coproc: Add initial draft for dummy remote processor coproc_xxx

Signed-off-by: Oleksandr Tyshchenko <olekstysh@gmail.com>

otyshchenko1 commented:
New patches arrived

aanisov commented on xen/arch/arm/coproc/plat/common.h:
```diff
@@ -1,6 +1,11 @@
 /*
  * xen/arch/arm/coproc/plat/common.h
  *
+ * Common platform header for all Remote processors
```
Why "remote processors"?

aanisov commented on xen/arch/arm/coproc/plat/common.c:
```diff
@@ -1,6 +1,12 @@
 /*
  * xen/arch/arm/coproc/plat/common.c
  *
+ * Common platform code for all Remote processors
```
Why "Remote processors"?

aanisov commented on xen/arch/arm/coproc/plat/coproc_xxx.c:
```diff
@@ -90,8 +89,27 @@ static int vcoproc_xxx_write(struct vcpu *v, mmio_info_t *info, register_t r,
             dabt.reg, r, offset, (uint32_t)mmio->addr);
 
+#if 1
     /* for debug purposes */
```
As I understand all the code here is for debug, so why do you stress it here and wrap with "#if 1"?
